src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/SystemTextJsonLanguageServer.cs

```diff
@@ -0,0 +1,107 @@
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+
+namespace Microsoft.CommonLanguageServerProtocol.Framework;
+
+internal abstract class SystemTextJsonLanguageServer<TRequestContext>(JsonRpc jsonRpc, JsonSerializerOptions options, ILspLogger logger) : AbstractLanguageServer<TRequestContext>(jsonRpc, logger)
+{
+    /// <summary>
+    /// JsonSerializer options used by streamjsonrpc (and for serializing / deserializing the requests to streamjsonrpc).
+    /// These options are specifically from the <see cref="StreamJsonRpc.SystemTextJsonFormatter"/> that added the exotic type converters.
+    /// </summary>
+    private readonly JsonSerializerOptions _jsonSerializerOptions = options;
+    protected override DelegatingEntryPoint CreateDelegatingEntryPoint(string method, IGrouping<string, RequestHandlerMetadata> handlersForMethod)
+    {
+        return new SystemTextJsonDelegatingEntryPoint(method, handlersForMethod, this);
+    }
+
+    protected virtual string GetLanguageForRequest(string methodName, JsonElement? parameters)
+    {
+        Logger.LogInformation($"Using default language handler for {methodName}");
+        return LanguageServerConstants.DefaultLanguageName;
+    }
+
+    private sealed class SystemTextJsonDelegatingEntryPoint(
+        string method,
+        IGrouping<string, RequestHandlerMetadata> handlersForMethod,
+        SystemTextJsonLanguageServer<TRequestContext> target) : DelegatingEntryPoint(method, handlersForMethod)
+    {
+        private static readonly MethodInfo s_parameterlessEntryPoint = typeof(SystemTextJsonDelegatingEntryPoint).GetMethod(nameof(SystemTextJsonDelegatingEntryPoint.ExecuteRequest0Async), BindingFlags.NonPublic | BindingFlags.Instance)!;
+        private static readonly MethodInfo s_entryPoint = typeof(SystemTextJsonDelegatingEntryPoint).GetMethod(nameof(SystemTextJsonDelegatingEntryPoint.ExecuteRequestAsync), BindingFlags.NonPublic | BindingFlags.Instance)!;
+
+        public override MethodInfo GetEntryPoint(bool hasParameter)
+        {
+            return hasParameter ? s_entryPoint : s_parameterlessEntryPoint;
+        }
+
+        /// <summary>
+        /// StreamJsonRpc entry point for handlers with no parameters.
+        /// Unlike Newtonsoft, we have to differentiate instead of using default parameters.
+        /// </summary>
+        private Task<JsonElement?> ExecuteRequest0Async(CancellationToken cancellationToken = default)
+        {
+            return ExecuteRequestAsync(null, cancellationToken);
+        }
+
+        /// <summary>
+        /// StreamJsonRpc entry point for handlers with parameters (and any response) type.
+        /// </summary>
+        private async Task<JsonElement?> ExecuteRequestAsync(JsonElement? request, CancellationToken cancellationToken = default)
+        {
+            var queue = target.GetRequestExecutionQueue();
```

Comment:
>target [](http://example.com/codeflow?start=24&length=6)

@CyrusNajmabadi 

Off topic: What's the general consensus on primary constructor usage? 

I like them when they are accessed during field initializations (outside methods). However, I strongly like being able to easily differentiate class vs local data within methods. I guess if I had my druthers, for any primary ctor arg needed in a method we'd still have the field declaration using the underscore syntax that sets itself to the primary constructor arg.

Reply:
currently no rules.  there are mixed feelings.  I personally have no issues with captures.  but i'm fairly used to lambdas, local functions, and JS/TS-style capture development.  I can understand the desire around fields though.

Maybe a general rule would be that it's ok for fairly simple types with few methods.  But once it gets complex enough, we assign to fields?